handlers/whatsapp/whatsapp.go

```diff
@@ -600,7 +608,127 @@ func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStat
+		elapsed := time.Now().Sub(start)
+		log := courier.NewChannelLogFromError("error reading the media id from redis", msg.Channel(), msg.ID(), elapsed, err)
+		logs = append(logs, log)
+		return "", logs, err
+	}
+
+	// check in failure cache
+	failureCacheKey := fmt.Sprintf(failureMediaCacheKeyPattern, msg.Channel().UUID().String())
+	if expiration, _ := getMediaExpirationFromCache(rc, failureCacheKey, mediaURL); expiration != nil {
+		return "", logs, errors.New("ignoring media that previously failed to upload")
+	}
+
+	// download media
+	req, err := http.NewRequest("GET", mediaURL, nil)
+	if err != nil {
+		return "", logs, err
+	}
+	res, err := utils.MakeHTTPRequest(req)
+	if err != nil {
+		elapsed := time.Now().Sub(start)
+		log := courier.NewChannelLogFromError("error downloading the media", msg.Channel(), msg.ID(), elapsed, err)
+		logs = append(logs, log)
+		return "", logs, err
+	}
+
+	// upload media to WhatsApp
+	baseURL := msg.Channel().StringConfigForKey(courier.ConfigBaseURL, "")
+	req, err = http.NewRequest("POST", baseURL + "/v1/media", bytes.NewReader(res.Body))
+	if err != nil {
+		return "", logs, err
+	}
+	setWhatsAppAuthHeader(&req.Header, msg.Channel())
+	req.Header.Add("Content-Type", mimeType)
+	res, err = utils.MakeHTTPRequest(req)
+	if err != nil {
+		// put in failure cache
+		setMediaExpirationInCache(rc, failureCacheKey, mediaURL, "", h.Server().Config().WhatsAppFailedMediaExpiration)
+
+		if res != nil {
+			err = errors.Wrap(err, string(res.Body))
+		}
+		elapsed := time.Now().Sub(start)
+		log := courier.NewChannelLogFromError("error uploading the media to WhatsApp", msg.Channel(), msg.ID(), elapsed, err)
+		logs = append(logs, log)
+		return "", logs, err
+	}
+
+	// take uploaded media id
+	mediaID, err := jsonparser.GetString(res.Body, "media", "[0]", "id")
+	if err != nil {
+		elapsed := time.Now().Sub(start)
+		log := courier.NewChannelLogFromError("error reading the media id from response", msg.Channel(), msg.ID(), elapsed, err)
+		logs = append(logs, log)
+		return "", logs, err
+	}
+
+	// put in cache
+	err = setMediaExpirationInCache(rc, cacheKey, mediaURL, mediaID, h.Server().Config().WhatsAppMediaExpiration)
+	if err != nil {
+		elapsed := time.Now().Sub(start)
+		log := courier.NewChannelLogFromError("error setting the media id to redis", msg.Channel(), msg.ID(), elapsed, err)
+		logs = append(logs, log)
+		return "", logs, err
+	}
+	return mediaID, logs, nil
+}
+
+type mediaExpiration struct {
+	ExpiresOn time.Time `json:"expires_on"`
+	MediaID   string    `json:"media_id,omitempty"`
+}
+
+func getMediaExpirationFromCache(rc redis.Conn, cacheKey, mediaURL string) (*mediaExpiration, error) {
+	expirationJSON, err := redis.String(rc.Do("HGET", cacheKey, mediaURL))
+	if err != nil {
+		return nil, err
+	}
+
+	expiration := &mediaExpiration{}
+	err = json.Unmarshal([]byte(expirationJSON), expiration)
+	if err != nil {
+		return nil, err
+	}
+
+	if expiration.ExpiresOn.Before(time.Now()) {
```

Comment:
What's going on here?  You are managing the expiration yourself? How are things going to get expired if  their expiration is only checked when they are fetched?

Reply:
Yes, I'm managing the expiration on my own because I switched the cache value type from string to hash and I can't set an expiration for a hash field. Key-value example of this implementation:

```
"whatsapp_media:a0b118cb-b010-43d8-a0b0-5483e125a812": {
  "https://link.to/image.jpg": "{\"expires_on\": \"2020-10-10T10:00:00+00:00\", \"media_id\": \"f043afd0-f0ae-4b9c-ab3d-696fb4c8cd68\"}",
  "https://link.to/video.mp4": "{\"expires_on\": \"2020-10-15T10:00:00+00:00\", \"media_id\": \"b4c8cd68-f0ae-4b9c-ab3d-696ff043afd0\"}"
}
```

Is there any way to expire a hash field automatically?

Reply:
That's just going to cause the cache to grow unbounded, things will never clear if they are only cleared when accessed after expiration. (simplest case of a piece of media being sent once and only once being an example)

Please use the new `rcache` stuff in `gocommon` > `v1.6.0`, that'll do what we want.

Reply:
Hey @nicpottier do you recommend some specific version of Redis? I'm facing the following error when I try to get a non-cached key: `NOSCRIPT No matching script. Please use EVAL.` (github.com/garyburd/redigo)

Reply:
Hrmm, that sounds weird, why don't you check in what you have and I'll take a look?

Reply:
Look:

![image](https://user-images.githubusercontent.com/24883556/96304187-0b2d5080-0fd2-11eb-9f04-e27ab06aebf1.png)

I'm using Redis 5.0.7.

Reply:
There's nothing specific to Redis versions no, Lua is available on all versions. I'm guessing there may be something subtle going on with redigo and the caching of the Lua script in that it thinks Redis has a compiled version stored but that isn't the case. Happy to take a look if you check your stuff in but that's where I would start looking.

Reply:
hmm, ok, thanks!

Reply:
Can you check in what you have so I can take a look?

Reply:
Hi @nicpottier, I still haven't been able to take a look at this, I can only confirm that I will check this week, sorry for the delay.

Reply:
@nicpottier as garyburd said [here](https://github.com/gomodule/redigo/issues/354#issuecomment-412750363), we have two redis packages, `github.com/gomodule/redigo` in gocommon and `github.com/garyburd/redigo` in courier.

Reply:
You can switch the courier dependency to `gomodule` in this PR. Should address the problem.